File: scilab/modules/linear_algebra/src/fortran/intzgeev.f
       subroutine intzgeev(fname)

c     [V,D]=spec(A)

      include 'stack.h'
      logical getrhsvar,createvar
      logical checklhs,checkrhs
      integer vfinite
      
      character fname*(*)
      character JOBVL, JOBVR
      complex*16 ZERO
      parameter ( ZERO=(0.0D0,0.0D0) )

      minrhs=1
      maxrhs=1
      minlhs=1
      maxlhs=2
c     
      if(.not.checkrhs(fname,minrhs,maxrhs)) return
      if(.not.checklhs(fname,minlhs,maxlhs)) return 

      if(.not.getrhsvar(1,'z', M, N, lA)) return
      if(M.ne.N) then
         err=1
         call error(20)
         return
      endif
      if(N.eq.0) then
         if(lhs.eq.1) then
            if(.not.createvar(2,'z',0,0,lD)) return
            lhsvar(1) = 2
            return
         else
            if(.not.createvar(2,'z', 0, 0, lD)) return
            if(.not.createvar(3,'d', 0, 0, lV)) return
            lhsvar(1) = 2
            lhsvar(2) = 3
            return
         endif
      elseif(N.eq.-1) then
         err=1
         call error(271)
         return
      endif  
      if(vfinite(2*M*N,zstk(lA)).eq.0) then
         err=1
         call error(264)
         return
      endif 
      if(lhs.eq.1) then
         if(.not.createvar(2,'z', N, 1, lD)) return
         k = 3              
      else if(lhs.eq.2) then
         if(.not.createvar(2,'z', N, N, lD)) return
         if(.not.createvar(3,'z', N, N, lVR)) return
         k = 4
      endif
      if(.not.createvar(k,'z', N, 1, lW)) return
      if(.not.createvar(k+1,'d', 2*N, 1, lRWORK)) return
      LWORKMIN = 2*N
      LWORK=maxvol(k+2,'z')
      if(LWORK.le.LWORKMIN) then
         err=2*(LWORK-LWORKMIN)
         call error(17)
         return
      endif
      if(.not.createvar(k+2,'z',1,LWORK,lDWORK)) return

      JOBVL = 'N'
      if(lhs.eq.1) then
         JOBVR = 'N'
      else
         JOBVR = 'V'
      endif 
      call ZGEEV( JOBVL, JOBVR, N, zstk(lA), N, zstk(lW), 
     $     zstk(lDWORK), N, zstk(lVR), N, zstk(lDWORK), LWORK,
     $     stk(lRWORK), INFO )
c     SUBROUTINE ZGEEV( JOBVL, JOBVR, N, A, LDA, W, VL, LDVL,
c     $     VR, LDVR, WORK, LWORK, RWORK, INFO )
      if(info.ne.0) then
         call error(24)
         return
      endif

      if(lhs.gt.1) then
         call ZLASET( 'F', N, N, ZERO, ZERO, zstk(lD), N ) 
c     SUBROUTINE ZLASET( UPLO, M, N, ALPHA, BETA, A, LDA )

         do 10 i = 1, N
            ii = i+(i-1)*N
            zstk(lD+ii-1) = zstk(lW+i-1)
 10      continue 
      else
         do 11 i = 1, N
            zstk(lD+i-1) = zstk(lW+i-1)
 11      continue
      endif
      
      if(lhs.eq.1) then
         lhsvar(1) = 2
      else
         lhsvar(1)=3
         lhsvar(2)=2
      endif
c     
      end

File: scilab/modules/linear_algebra/src/c/intzgeev.c
/*
 * Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
 * Copyright (C) 2008 - INRIA - Michaël Baudin
 * 
 * This file must be used under the terms of the CeCILL.
 * This source file is licensed as described in the file COPYING, which
 * you should have received as part of this distribution.  The terms
 * are also available at    
 * http://www.cecill.info/licences/Licence_CeCILL_V2-en.txt
 *
 */
#include "stack-c.h" 
#include "elementary_functions.h" 
#include "do_error_number.h"
#include "MALLOC.h"
#include "vfinite.h"
#include "assembleEigenvectors.h"

extern int C2F(dlaset)();
extern int C2F(zgeev)();

//
// intzgeev --
//   Interface to LAPACK's ZGEEV
//   Computes the eigenvalues and, if required, the eigenvectors of a complex un-symetric matrix.
//   Possible uses :
//   * With 1 LHS :
//     eigenvalues=spec(A)
//   where 
//     A : symetric, square matrix of size NxN
//     eigenvalues : matrix of size Nx1, type complex
//   * With 2 LHS :
//     [eigenvectors,eigenvalues]=spec(A)
//   where 
//     A : square matrix of size NxN
//     eigenvalues : matrix of size NxN with eigenvalues as diagonal terms, type complex
//     eigenvectors : matrix of size NxN, type complex
//
int intzgeev(char *fname)
{	
	int lA = 0;
	int totalsize;
	int iRows = 0;
	int iCols = 0;
	int ONE = 1;
	int iWorkSize;
	int INFO;

	char JOBVL;
	char JOBVR;

	double *pdblDataReal = NULL;
	double *pdblDataImg = NULL;
	double *pdblFinalEigenvaluesReal = NULL; //SCILAB return Var
	double *pdblFinalEigenvaluesImg = NULL; //SCILAB return Var
	double *pdblFinalEigenvectorsReal = NULL; //SCILAB return Var
	double *pdblFinalEigenvectorsImg = NULL; //SCILAB return Var
	doublecomplex *pdblData	= NULL;
	doublecomplex *pdblEigenValues= NULL; //return by LAPACK
	doublecomplex *pdblWork = NULL; // Used by LAPACK
	doublecomplex *pdblRWork = NULL; // Used by LAPACK
	doublecomplex * pdblLeftvectors = NULL; // Used by LAPACK
	doublecomplex * pdblRightvectors = NULL; // Used by LAPACK

	CheckRhs(1,1) ;
	CheckLhs(1,2) ;

	GetRhsVarMatrixComplex(1, &iRows, &iCols, &pdblDataReal, &pdblDataImg);
	totalsize = iRows * iCols;
	pdblData = oGetDoubleComplexFromPointer(pdblDataReal, pdblDataImg, totalsize);
	
	if (iRows!=iCols)
	{
		SciError(20);
		return 0;
	}
	if (iCols==0)
	{
		if (Lhs==1)
		{
			int lD;
			CreateVar(2,MATRIX_OF_COMPLEX_DATATYPE,&iCols,&iCols,&lD);
			LhsVar(1) = 2;
			return 0;
		}
		else if (Lhs==2)
		{
			int lD;
			int lV;
			CreateVar(2,MATRIX_OF_COMPLEX_DATATYPE,&iCols,&iCols,&lD);
			CreateVar(3,MATRIX_OF_DOUBLE_DATATYPE,&iCols,&iCols,&lV);
			LhsVar(1) = 2;
			LhsVar(2) = 3;
			return 0;
		}
	}
	if (C2F(vfiniteComplex)(&totalsize,pdblData)==0)
	{
		SciError(264);
		return 0;
	}
	if (Lhs==1)
	{
		iAllocMatrixOfDoubleComplex(2, iCols, ONE, &pdblFinalEigenvaluesReal, &pdblFinalEigenvaluesImg);
	}
	else
	{
		iAllocMatrixOfDoubleComplex(2, iCols, iCols, &pdblFinalEigenvaluesReal, &pdblFinalEigenvaluesImg);
		iAllocMatrixOfDoubleComplex(3, iCols, iCols, &pdblFinalEigenvectorsReal, &pdblFinalEigenvectorsImg);
		pdblRightvectors = (doublecomplex*)MALLOC(sizeof(doublecomplex) * totalsize);
	}

	pdblEigenValues = (doublecomplex*)MALLOC(sizeof(doublecomplex) * iCols);

	iWorkSize = Max(1,2*iCols);
	pdblWork = (doublecomplex*)MALLOC(sizeof(doublecomplex) * iWorkSize);
	pdblRWork = (doublecomplex*)MALLOC(sizeof(doublecomplex) * 2*iCols);

	JOBVL = 'N';
	if (Lhs==1)
	{
		JOBVR = 'N'; // Compute eigenvalues only;
	}
	else
	{
		JOBVR = 'V'; // Compute eigenvalues and eigenvectors.
	}
	C2F(zgeev)( &JOBVL, &JOBVR, &iCols, pdblData, &iCols, pdblEigenValues, 
		pdblLeftvectors, &iCols, pdblRightvectors, &iCols, pdblWork, &iWorkSize,
		pdblRWork, &INFO );
//     SUBROUTINE ZGEEV( JOBVL, JOBVR, N, A, LDA, W, VL, LDVL,
//     $     VR, LDVR, WORK, LWORK, RWORK, INFO )
	FREE(pdblWork);
	FREE(pdblRWork);
	if (INFO!=0)
	{
		SciError(24);
	}
	if(Lhs==2)
	{
		// Transfert eigenvalues
		assembleComplexEigenvaluesFromDoubleComplexPointer(iRows, pdblEigenValues, 
			pdblFinalEigenvaluesReal, pdblFinalEigenvaluesImg);
		// Transfert eigenvectors from doublecomplex to real and imaginary parts
		vGetPointerFromDoubleComplex(pdblRightvectors, totalsize, pdblFinalEigenvectorsReal, pdblFinalEigenvectorsImg);
	}
	else
	{
		// Transfert eigenvalues from doublecomplex to real and imaginary parts
		vGetPointerFromDoubleComplex(pdblEigenValues, iCols, pdblFinalEigenvaluesReal, pdblFinalEigenvaluesImg);
	}
	if(Lhs==1)
	{
		LhsVar(1)=2;
	}
	else
	{
		LhsVar(1)=3;
		LhsVar(2)=2;
	}
	FREE(pdblEigenValues);
	if(Lhs==2)
	{
		FREE(pdblRightvectors);
	}
	vFreeDoubleComplexFromPointer(pdblData);
	return 0;
}

